File: src/main/java/de/pflugradts/passbird/domain/model/namespace/Namespaces.java
package de.pflugradts.passbird.domain.model.namespace;

import de.pflugradts.passbird.domain.model.transfer.Bytes;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.CAPACITY;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.DEFAULT;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.FIRST_NAMESPACE;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.LAST_NAMESPACE;

public class Namespaces {

    private final List<Optional<Namespace>> namespacesList = new ArrayList<>();
    private NamespaceSlot currentNamespace = DEFAULT;

    public void reset() {
        namespacesList.clear();
    }

    private List<Optional<Namespace>> getNamespaces() {
        return namespacesList;
    }

    public void populate(final List<Bytes> namespaceBytes) {
        if (getNamespaces().isEmpty()) {
            if (namespaceBytes.size() == CAPACITY) {
                IntStream.range(FIRST_NAMESPACE, LAST_NAMESPACE + 1).forEach(index ->
                    getNamespaces().add(namespaceBytes.get(index - 1).isEmpty()
                        ? Optional.empty()
                        : Optional.of(Namespace.create(namespaceBytes.get(index - 1), NamespaceSlot.at(index)))));
            } else {
                IntStream.range(FIRST_NAMESPACE, LAST_NAMESPACE + 1).forEach(x -> getNamespaces().add(Optional.empty()));
            }
        }
    }

    public void deploy(final Bytes namespaceBytes, final NamespaceSlot namespaceSlot) {
        getNamespaces().set(namespaceSlot.index() - 1, Optional.of(Namespace.create(namespaceBytes, namespaceSlot)));
    }

    public Optional<Namespace> atSlot(final NamespaceSlot namespaceSlot) {
        return namespaceSlot == DEFAULT
            ? Optional.of(Namespace.DEFAULT)
            : getNamespaces().get(namespaceSlot.index() - 1);
    }

    public Stream<Optional<Namespace>> all() {
        return getNamespaces().stream();
    }

    public Namespace getCurrentNamespace() {
        return atSlot(currentNamespace).orElse(Namespace.DEFAULT);
    }

    public void updateCurrentNamespace(final NamespaceSlot namespaceSlot) {
        if (atSlot(namespaceSlot).isPresent()) {
            currentNamespace = namespaceSlot;
        }
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/namespace/Namespaces.kt
package de.pflugradts.passbird.domain.model.namespace

import de.pflugradts.passbird.domain.model.namespace.Namespace.Companion.DEFAULT
import de.pflugradts.passbird.domain.model.namespace.Namespace.Companion.createNamespace
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.Companion.CAPACITY
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.Companion.FIRST_NAMESPACE
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.Companion.LAST_NAMESPACE
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.Companion.at
import de.pflugradts.passbird.domain.model.transfer.Bytes
import java.util.Collections
import java.util.Optional

class Namespaces {
    private val namespaces = Collections.nCopies(CAPACITY, Optional.empty<Namespace>()).toMutableList()
    private var currentNamespace = NamespaceSlot.DEFAULT

    fun reset() = (0..<CAPACITY).forEach { namespaces[it] = Optional.empty() }

    fun populate(namespaceBytes: List<Bytes>) {
        if (namespaces.find { it.isPresent } == null && namespaceBytes.size == NamespaceSlot.CAPACITY) {
            (FIRST_NAMESPACE..LAST_NAMESPACE).forEach {
                if (namespaceBytes[it - 1].isNotEmpty) {
                    namespaces[it - 1] = Optional.of(createNamespace(namespaceBytes[it - 1], at(it)))
                }
            }
        }
    }

    fun deploy(namespaceBytes: Bytes, namespaceSlot: NamespaceSlot) {
        namespaces[namespaceSlot.index() - 1] = Optional.of(createNamespace(namespaceBytes, namespaceSlot))
    }
    fun atSlot(namespaceSlot: NamespaceSlot): Optional<Namespace> =
        if (namespaceSlot === NamespaceSlot.DEFAULT) Optional.of(DEFAULT) else namespaces[namespaceSlot.index() - 1]
    fun all() = namespaces.stream()
    fun getCurrentNamespace() = atSlot(currentNamespace).orElse(DEFAULT)
    fun updateCurrentNamespace(namespaceSlot: NamespaceSlot) { if (atSlot(namespaceSlot).isPresent) { currentNamespace = namespaceSlot } }
}
